futures-util/src/io/read_line.rs

```diff
@@ -0,0 +1,55 @@
+use futures_core::future::Future;
+use futures_core::task::{Context, Poll};
+use futures_io::AsyncBufRead;
+use std::io;
+use std::pin::Pin;
+use std::str;
+use super::read_until::read_until_internal;
+
+/// Future for the [`read_line`](super::AsyncBufReadExt::read_line) method.
+#[derive(Debug)]
+pub struct ReadLine<'a, R: ?Sized + Unpin> {
+    reader: &'a mut R,
+    buf: &'a mut String,
+    read: usize,
+}
+
+impl<R: ?Sized + Unpin> Unpin for ReadLine<'_, R> {}
+
+impl<'a, R: AsyncBufRead + ?Sized + Unpin> ReadLine<'a, R> {
+    pub(super) fn new(reader: &'a mut R, buf: &'a mut String) -> Self {
+        Self { reader, buf, read: 0 }
+    }
+}
+
+struct Guard<'a> { buf: &'a mut Vec<u8>, len: usize }
+
+impl Drop for Guard<'_> {
+    fn drop(&mut self) {
+        unsafe { self.buf.set_len(self.len); }
+    }
+}
+
+impl<R: AsyncBufRead + ?Sized + Unpin> Future for ReadLine<'_, R> {
+    type Output = io::Result<usize>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let Self { reader, buf, read } = &mut *self;
+        unsafe {
+            // safety: https://doc.rust-lang.org/src/std/io/mod.rs.html#310
+            let mut g = Guard { len: buf.len(), buf: buf.as_mut_vec() };
+            let poll = read_until_internal(Pin::new(reader), b'\n', g.buf, read, cx);
+            if str::from_utf8(&g.buf[g.len..]).is_err() {
```

Comment:
`read_until_internal` could have read the first bytes of a multi-byte character then returned `Pending` at this point. I don’t think you can really look at the content of the buffer until you get a `Ready`.

Reply:
Hmm, I think you're right.

AFAIK the problem with the previous implementation was that `g.len` was updated when `Pending` was called. So perhaps it can resolve it by recording the first `buf.len()` and getting `g.len` from the field of `ReadLine` (i.e., `self.len`). And when `Ready` returned, it updates `self.len` and `g.len` only when it is a valid string, and if it is an invalid string, it just drops `g`. Also, when `Pending` is returned, update only `g.len`.

Reply:
An alternative would be to add an extra `bytes: Vec<u8>` field to the struct, do `bytes: mem::replace(buf.as_mut_vec(), Vec::new())` in the constructor, then once you receive `Ready` validate the bytes and `mem::swap(buf.as_mut_vec(), bytes)`. That way you avoid the whole per-poll guard thing.

Reply:
Oh, that's the safer way. I like that, thanks.